pkg/extensions/sync/remote.go

```diff
@@ -5,160 +5,348 @@ package sync
+
+			continue
+		}
+
+		return repoList.Repositories, nil
+	}
+
+	return []string{}, err
+}
 
-	_, _, _, err := registry.client.MakeGetRequest(ctx, &catalog, "application/json", //nolint: dogsled
-		constants.RoutePrefix, constants.ExtCatalogPrefix)
+func (registry *RemoteRegistry) GetImageReference(repo, reference string) (ref.Ref, error) {
+	digest, ok := parseReference(reference)
+
+	var imageRefPath string
+	if ok {
+		imageRefPath = fmt.Sprintf("%s/%s@%s", registry.host, repo, digest.String())
+	} else {
+		// is tag
+		imageRefPath = fmt.Sprintf("%s/%s:%s", registry.host, repo, reference)
+	}
+
+	imageRef, err := ref.New(imageRefPath)
 	if err != nil {
-		return []string{}, err
+		return ref.Ref{}, err
 	}
 
-	return catalog.Repositories, nil
+	return imageRef, nil
 }
 
-func (registry *RemoteRegistry) GetDockerRemoteRepo(repo string) string {
-	dockerNamespace := "library"
-	dockerRegistry := "docker.io"
+func (registry *RemoteRegistry) GetOCIManifest(ctx context.Context, repo, reference string,
+) ([]byte, ispec.Descriptor, bool, error) {
+	var isConverted bool
+
+	var buf []byte
 
-	remoteHost := registry.client.GetHostname()
+	var desc ispec.Descriptor
 
-	repoRef, err := parseRepositoryReference(fmt.Sprintf("%s/%s", remoteHost, repo))
+	imageReference, err := registry.GetImageReference(repo, reference)
 	if err != nil {
-		return repo
+		return nil, ispec.Descriptor{}, false, err
 	}
 
-	if !strings.Contains(repo, dockerNamespace) &&
-		strings.Contains(repoRef.String(), dockerNamespace) &&
-		strings.Contains(repoRef.String(), dockerRegistry) {
-		return fmt.Sprintf("%s/%s", dockerNamespace, repo)
+	/// check what error it gives when not found
+	man, err := registry.client.ManifestGet(ctx, imageReference)
+	if err != nil {
+		/* public registries may return 401 for image not found
+		they will try to check private registries as a fallback => 401 */
+		if errors.Is(err, errs.ErrHTTPUnauthorized) {
+			registry.log.Info().Str("errorType", common.TypeOf(err)).
+				Str("repository", repo).Str("reference", reference).
+				Err(err).Msg("failed to get manifest: unauthorized")
+
+			return nil, ispec.Descriptor{}, false, zerr.ErrUnauthorizedAccess
+		} else if errors.Is(err, errs.ErrNotFound) {
+			registry.log.Info().Str("errorType", common.TypeOf(err)).
+				Str("repository", repo).Str("reference", reference).
+				Err(err).Msg("failed to find manifest")
+
+			return nil, ispec.Descriptor{}, false, zerr.ErrManifestNotFound
+		}
+
+		return nil, ispec.Descriptor{}, false, err
+	}
+
+	switch man.GetDescriptor().MediaType {
+	case mediatype.Docker2Manifest:
+		buf, desc, err = convertDockerManifestToOCI(ctx, man, imageReference, registry.client)
+		isConverted = true
+	case mediatype.Docker2ManifestList:
+		buf, desc, err = convertDockerListToOCI(ctx, man, imageReference, registry.client)
+		isConverted = true
+	case mediatype.OCI1Manifest, mediatype.OCI1ManifestList:
```

Comment:
Off topic: let's make sure when we sync we don't unmarshall/marshal the manifest when writing to the destination.

Currently there's an open bug that sync reorders the fields in the image manifest, resulting in a different digest, and broken referrers.

Reply:
This is only for docker images we don't do any marshal unmarshal for oci.

For docker images I guess it's normal that referrers are lost in process of converting.